unstructured/ingest/connector/fsspec.py

```diff
@@ -209,16 +220,104 @@ def _list_files(self):
                 if v.get("size") > 0
             ]
 
+    def _decompress_file(self, filename: str) -> str:
+        head, tail = os.path.split(filename)
+        if any(filename.endswith(ext) for ext in zip_file_extensions):
+            for ext in zip_file_extensions:
+                if tail.endswith(ext):
+                    tail = tail[: -(len(ext))]
+                    break
+            path = os.path.join(head, f"{tail}-zip-uncompressed")
+            logger.info(f"extracting zip {filename} -> {path}")
+            with zipfile.ZipFile(filename) as zfile:
+                zfile.extractall(path=path)
+            return path
+        elif any(filename.endswith(ext) for ext in tar_file_extensions):
+            for ext in tar_file_extensions:
+                if tail.endswith(ext):
+                    tail = tail[: -(len(ext))]
+                    break
+
+            path = os.path.join(head, f"{tail}-tar-uncompressed")
+            logger.info(f"extracting tar {filename} -> {path}")
+            with tarfile.open(filename, "r:gz") as tfile:
+                tfile.extractall(path=path)
+            return path
+        else:
+            raise ValueError(
+                "filename {} not a recognized compressed extension: {}".format(
+                    filename,
+                    ", ".join(zip_file_extensions + tar_file_extensions),
+                ),
+            )
```

Comment:
make this re-usable/common/accessible to other connectors that might support compressed files in the future?

Reply:
Can you write a unit test for this? Just to make sure it does what we all expect.